internal/terraform/node_resource_validate.go

```diff
@@ -551,6 +520,187 @@ func (n *NodeValidatableResource) validateCheckRules(ctx EvalContext, config *co
 	return diags
 }
 
+// validateImportTargets checks that the import block expressions are valid.
+func (n *NodeValidatableResource) validateImportTargets(ctx EvalContext) tfdiags.Diagnostics {
+	var diags tfdiags.Diagnostics
+
+	if len(n.importTargets) == 0 {
+		return diags
+	}
+
+	diags = diags.Append(n.validateConfigGen(ctx))
+
+	// Import blocks are only valid within the root module, and must be
+	// evaluated within that context
+	ctx = evalContextForModuleInstance(ctx, addrs.RootModuleInstance)
+
+	for _, imp := range n.importTargets {
+		if imp.Config == nil {
+			// if we have a legacy addr, it will be supplied on the command line so
+			// there is nothing to check now and we need to wait for plan.
+			continue
+		}
+
+		diags = diags.Append(validateImportSelfRef(n.Addr.Resource, imp.Config.ID))
+		if diags.HasErrors() {
+			return diags
+		}
+
+		// Resource config might be nil here since we are also validating config generation.
+		expanded := n.Config != nil && (n.Config.ForEach != nil || n.Config.Count != nil)
+
+		if imp.Config.ForEach != nil {
+			if !expanded {
+				diags = diags.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Use of import for_each in an invalid context",
+					Detail:   "Use of for_each in import requires a resource using count or for_each.",
+					// FIXME: minor issue, but this points to the for_each expression rather than for_each itself.
+					Subject: imp.Config.ForEach.Range().Ptr(),
+				})
+			}
+
+			forEachData, _, forEachDiags := newForEachEvaluator(imp.Config.ForEach, ctx, true).ImportValues()
+			diags = diags.Append(forEachDiags)
+			if forEachDiags.HasErrors() {
+				return diags
+			}
+
+			for _, keyData := range forEachData {
+				to, evalDiags := evalImportToExpression(imp.Config.To, keyData)
+				diags = diags.Append(evalDiags)
+				if diags.HasErrors() {
+					return diags
+				}
+				diags = diags.Append(n.validateImportTargetExpansion(to, imp.Config.To))
```

Comment:
is it worth also adding this new validation to the equivalent part of the plan? It only matters in situations where the expansion is unknown during the validation, and I guess we might get a different error message about the resource being missing or something.

Just wondering if it's worth being consistent?

Reply:
That's a good point, the plan error will just say the resource is missing, and these errors probably more helpful. I initially wanted to do this all statically (which I think is nearly possible except when the use has `any` typed variables involved) but carrying the type info forward to determine what `each.key` is is not so easy.

Reply:
We can often still get the type of the `for_each` expression when it's unknown, and we have the `To` traversal. It would be easy to validate if the `To` expression only allowed indexing with `each.key` but in reality it can use anything from `each`, so we need to find and extract the resource index key expression and apply that path to the `for_each` type to determine if we have a numeric or string index. We're so close to being able to statically verify this, but the required code just felt too big and cumbersome to throw in for these tiny edge cases and still not cover the cases where the type is unknown. 

Instead added the same validation checks into the plan node too for the better error messages.